### PR TITLE
Implement record factory for database

### PR DESCRIPTION
**Pull request by oskarcieslik**
This implements #27

### CODE REVIEW

File: src/database.js [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
+import { recordFactory } from './recordFactory';
 
 export class Database {
   constructor() {
     this.setInitialState();
   }
 
   all(collectionName) {
-    return this.filter(collectionName);
+    return this.filter(collectionName, {});
```
Comment (zzarcon) [inline src/database.js]: 👍

File: src/database.js [hunk 2]
```diff
@@ -28,7 +29,7 @@ export class Database {
       records.push(this.decorateRecord(collectionName, record));
     }
 
-    this.store[collectionName] = records;
+    this._pushToStore(collectionName, records);
```
Comment (zzarcon) [inline src/database.js]: I don't like underscored methods, I think we need to discuss that ^^

File: src/database.js [hunk 3]
```diff
@@ -83,4 +84,14 @@ export class Database {
...
   }
+
+  _update(collectionName, record) {
+    const originalRecord = this.find(collectionName, {id: record.id});
+    Object.assign(originalRecord, record);
+  }
+
+  _pushToStore(collectionName, records) {
+    this.store[collectionName] = records.map(record =>
+      recordFactory(record, collectionName, this));
```
Comment (zzarcon) [inline src/database.js]: 👍

File: src/database.js (hunk 2 above)
Comment (oskarcieslik) [inline src/database.js]: It's just a reminder that this method should only be used as database internally, it still can be used outside of it but it tells user that he probably shouldn't do this manually.

Comment (oskarcieslik) [inline src/database.js]: Oh, you said you don't like them, I have read this as "i dont understand" ;p

File: src/database.js [hunk 4]
```diff
@@ -83,4 +84,14 @@ export class Database {
 
     return id;
   }
+
+  _update(collectionName, record) {
+    const originalRecord = this.find(collectionName, {id: record.id});
```
Comment (zzarcon) [inline src/database.js]: This should use `let` right?

File: src/database.js [hunk 5]
```diff
@@ -83,4 +84,14 @@ export class Database {
...
+    Object.assign(originalRecord, record);
+  }
```
Comment (zzarcon) [inline src/database.js]: Should update return the record?

File: src/database.js (hunk 4 above)
Comment (oskarcieslik) [inline src/database.js]: Nope, const in ES6 means that reference to object that this variable is pointing to can't change, not properties of this object itself :)

Comment (oskarcieslik) [inline src/database.js]: https://mathiasbynens.be/notes/es6-const

File: src/database.js (hunk 5 above)
Comment (oskarcieslik) [inline src/database.js]: Hmm, I suppose in some scenarios that might be useful, so yes :)

File: src/database.js (hunk 4 above)
Comment (zzarcon) [inline src/database.js]: Right!

File: tests/specs/database_spec.js [hunk 6]
```diff
@@ -167,6 +168,24 @@ export const databaseSpec = () => {
     assert.end();
   });
 
+  test('DB # record # save', assert => {
```
Comment (zzarcon) [inline tests/specs/database_spec.js]: I don't understand this test

File: src/database.js (hunk 1 above)
Comment (zzarcon) [inline src/database.js]: Actually not... This should be `this.store[collectionName]`

File: tests/specs/database_spec.js [hunk 7]
```diff
@@ -205,4 +225,80 @@ export const databaseSpec = () => {
 
     assert.end();
   });
+
+  test('recordFactory # _update', assert => {
```
Comment (zzarcon) [inline tests/specs/database_spec.js]: Should we test private methods?

File: tests/specs/database_spec.js [hunk 8]
```diff
@@ -205,4 +225,80 @@ export const databaseSpec = () => {
...
+    assert.end();
+  });
+
+  test('recordFactory # save', assert => {
+    const db = new Database();
+
+    db.register('user', userFactory);
+    db.create('user', 2);
+
+    const user1 = db.store.user[0];
```
Comment (zzarcon) [inline tests/specs/database_spec.js]: I think we shouldn't access the store directly

Comment (oskarcieslik) [inline tests/specs/database_spec.js]: k

File: tests/specs/database_spec.js (hunk 6 above)
Comment (oskarcieslik) [inline tests/specs/database_spec.js]: I updated it to take use of save()

File: tests/specs/database_spec.js [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import _ from 'lodash';
 import test from 'tape';
 import { Database } from '../../src/kakapo';
+import { recordFactory } from '../../src/recordFactory';
```
Comment (zzarcon) [inline tests/specs/database_spec.js]: this is not used

Comment (oskarcieslik) [inline tests/specs/database_spec.js]: right 👍

**Pull request merged**